### PR TITLE
Add BertLayer

### PR DESCRIPTION
**Pull request by fredblain**
This PR refers to https://github.com/RSE-Sheffield/deepquest/issues/43.

It adds a new Layer, named BertLayer, which aims at extracting word/token contextualised embedding using a pre-trained BERT Base Multilingual Cased model (from tensorflow-hub).
Based on a couple of experiments, the results seem convincing using this layer in a BiRNN sent-level QE model.

### CODE REVIEW

**Review by davidwyld — CHANGES_REQUESTED** [review]
As it is, a simple check out of this branch runs but without BERT tokenization.
I changed `TOKENIZATION_METHOD` to `'bert'` but this results in an error:
```
Exception has occurred: Exception
Tokenization procedure "bert" is not implemented.
  File "/home/ac1dwi/impactwork/deepquest/deepquest/data_engine/dataset.py", line 339, in preprocessText
    raise Exception('Tokenization procedure "' + tokenization + '" is not implemented.')
  File "/home/ac1dwi/impactwork/deepquest/deepquest/data_engine/dataset.py", line 225, in setInput
    bpe_codes=bpe_codes, separator=separator, use_unk_class=use_unk_class, bert_hub_module_handle=bert_hub_module_handle)
  File "/home/ac1dwi/impactwork/deepquest/deepquest/data_engine/prepare_data.py", line 458, in build_dataset
    bpe_codes=params.get('BPE_CODES_PATH', None))
  File "/home/ac1dwi/impactwork/deepquest/deepquest/train.py", line 91, in train_model
    dataset = build_dataset(params)
  File "/home/ac1dwi/impactwork/deepquest/deepquest/train.py", line 502, in main
    weights_path=parameters.get('PRED_WEIGHTS', None))
  File "/home/ac1dwi/impactwork/deepquest/deepquest/__main__.py", line 24, in train
    train(parameters)
  File "/home/ac1dwi/impactwork/deepquest/deepquest/__main__.py", line 105, in main
    args.func(args)
  File "/home/ac1dwi/impactwork/deepquest/deepquest/dq.py", line 4, in <module>
    sys.exit(main())
```

what other changes need to be made, to the config file for instance, to make this work?

**Comment by fredblain** [comment]
Your `TOKENIZATION_METHOD` was not set properly, instead use:
```
TOKENIZATION_METHOD: 'tokenize_bert'
```
by default, it is set to `tokenize_none`, not `none`.

What is missing is what we discussed last Friday: the possibility to add a new option/parameters. At the moment, both `update_parameters` and `check_params` functions are imported from dependencies.

**Comment by fredblain** [comment]
I pause this PR for now, as I need to check something.

**Comment by fredblain** [comment]
OK, @davidwilby and @drj11, I reopen this PR after refactoring `qe_model/utils.py`.
Please review and merge it if all is good for you.

**Comment by davidwyld** [comment]
@fredblain I'm having problems running a training task with bert tokenization for the test sentence dataset with a laptop gpu. Which CUDA, cuDNN and tensorflow/tensorflow-gpu versions have you been using when running this? I'm getting out of memory errors when running with bert tokenization, but not when using `tokenize_none`

**Comment by fredblain** [comment]
That's because you don't have enough memory on your laptop GPU, how much memory do you have? Running BERT requires resources, e.g. it takes about ~11Go RAM on a GeForce RTX 2080 Ti. Try to run it on your CPU rather than GPU, it will take longer, but it will use your laptop's RAM memory instead. If you have at least 16Go RAM on your laptop, it should run (it does on mine). If you have stuff already running, it is likely that it will swap, but it's for testing and you can clean the swap afterwards.
I don't know your configuration, so keep an eye on it so you don't crash your laptop. Can you try on Sharc otherwise?

`tokenize_none` means no tokenization, it does not require anything fancy, therefore that's normal it works fine on your GPU/CPU. `tokenize_bert` means that the data should be tokenized and processed for being used later on witth the BERT layer for token-level representation extraction.

**Comment by davidwyld** [comment]
Great, thanks Fred. FYI this ran successfully (although slowly) for the sentence test dataset.

**Review by davidwyld — APPROVED** [review]
Have run a training run on a sentence test dataset on CPU and it runs. I don't know what more to look for at this stage.
Obviously, using bert tokenization takes considerably more resources than without, but it runs.

Nice on Fred!

**Comment by fredblain** [comment]
@davidwilby, if @drj11 has no remark, we can merge.

File: deepquest/utils/bert_layer.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import tensorflow as tf
+import tensorflow_hub as hub
+from tensorflow.keras import backend as K
+
+from keras.engine.base_layer import Layer
+
+
+logging.basicConfig(level=logging.INFO,
+                    format='[%(asctime)s] %(message)s', datefmt='%d/%m/%Y %H:%M:%S')
+logger = logging.getLogger(__name__)
```
Comment (drj11) [inline deepquest/utils/bert_layer.py]: isn't this whole thing in layers.py now?

**Review by drj11 — APPROVED** [review]

File: deepquest/utils/bert_layer.py (hunk 1 above)
Comment (fredblain) [reply deepquest/utils/bert_layer.py]: yes, well spotted @drj11.

Comment (drj11) [reply deepquest/utils/bert_layer.py]: so delete that and assuming it still runs we're good to go.

**Comment by drj11** [comment]
I rally like that you've integrated BERT. :)

File: deepquest/utils/bert_layer.py (hunk 1 above)
Comment (fredblain) [reply deepquest/utils/bert_layer.py]: Removed. Travis OK. Resolving.

**Pull request merged**